### PR TITLE
Glyph Fixes

### PR DESCRIPTION
**Pull request by Haydelj**
Resolves issue #1468 and #1455. Also enables a few features in show field glyph.

### CODE REVIEW

File: src/Core/GeometryPrimitives/Vector.h [hunk 1]
```diff
@@ -384,6 +385,17 @@ inline Vector Cross(const Vector& v1, const Vector& v2)
 		v1.d_[0]*v2.d_[1]-v1.d_[1]*v2.d_[0]);
 }
 
+//returns a unit vector
+inline Vector Vector::getArbitraryTangent()
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Vector.h]:
```suggestion
inline Vector Vector::getArbitraryTangent() const
```

**Comment by dcwhite** [comment]
@Haydelj Nice! I like the co-author bit. I'll add the rest of my suggestions now.

File: src/Core/GeometryPrimitives/Vector.h [hunk 2]
```diff
@@ -384,6 +385,17 @@ inline Vector Cross(const Vector& v1, const Vector& v2)
 		v1.d_[0]*v2.d_[1]-v1.d_[1]*v2.d_[0]);
 }
 
+//returns a unit vector
+inline Vector Vector::getArbitraryTangent() const
+{
+  Vector normalized = this->normal();
+  
+  if(abs(normalized.x()) < 0.707)
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Vector.h]:
```suggestion
  if(abs(normalized.x()) < THRESHOLD)
```

File: src/Core/GeometryPrimitives/Vector.h [hunk 3]
```diff
@@ -384,6 +385,17 @@ inline Vector Cross(const Vector& v1, const Vector& v2)
 		v1.d_[0]*v2.d_[1]-v1.d_[1]*v2.d_[0]);
 }
 
+//returns a unit vector
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Vector.h]:
```suggestion
//returns a unit vector
static const double THRESHOLD = std::sqrt(2.0)/2;
```

File: src/Core/GeometryPrimitives/Vector.h [hunk 4]
```diff
@@ -384,6 +385,17 @@ inline Vector Cross(const Vector& v1, const Vector& v2)
...
+inline Vector Vector::getArbitraryTangent() const
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Vector.h]:
```suggestion
static const Vector i(1,0,0);
static const Vector j(0,1,0);
inline Vector Vector::getArbitraryTangent() const
```

File: src/Core/GeometryPrimitives/Vector.h [hunk 5]
```diff
@@ -384,6 +385,17 @@ inline Vector Cross(const Vector& v1, const Vector& v2)
...
+{
+  Vector normalized = this->normal();
+  
+  if(abs(normalized.x()) < 0.707)
+     return (Vector(1,0,0) - normalized * normalized.x()).normal();
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Vector.h]:
```suggestion
     return (i - normalized * normalized.x()).normal();
```

File: src/Core/GeometryPrimitives/Vector.h [hunk 6]
```diff
@@ -384,6 +385,17 @@ inline Vector Cross(const Vector& v1, const Vector& v2)
...
+  
+  return (Vector(0,1,0) - normalized * normalized.y()).normal();
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Vector.h]:
```suggestion
  return (j - normalized * normalized.y()).normal();
```

File: src/Modules/Visualization/ShowFieldGlyphs.cc [hunk 7]
```diff
@@ -342,161 +404,195 @@ void GlyphBuilder::renderVectors(
...
-  // Render cell data
-  if (!finfo.is_linear())
+  
+  bool normalizeGlyphs = state->getValue(ShowFieldGlyphs::NormalizeGlyphs).toBool();
+  bool renderBidirectionaly = state->getValue(ShowFieldGlyphs::RenderBidirectionaly).toBool();
+  bool renderGlphysBellowThreshold = state->getValue(ShowFieldGlyphs::RenderGlyphsBellowThreshold).toBool();
+  float threshold = state->getValue(ShowFieldGlyphs::Threshold).toDouble();
+
+  //sets feild location for consant feild data 1: node centered 2: edge centered 3: face centered 4: cell centered
+  int feildLocation = finfo.is_point()*1 + finfo.is_line()*2 + finfo.is_surface()*3 + finfo.is_volume()*4;
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowFieldGlyphs.cc]:
```suggestion
  int fieldLocation = finfo.is_point()*1 + finfo.is_line()*2 + finfo.is_surface()*3 + finfo.is_volume()*4;
```

File: src/Modules/Visualization/ShowFieldGlyphs.cc [hunk 8]
```diff
@@ -342,161 +404,195 @@ void GlyphBuilder::renderVectors(
...
-  bool done = false;
-  // Render cell data
-  if (!finfo.is_linear())
+  
+  bool normalizeGlyphs = state->getValue(ShowFieldGlyphs::NormalizeGlyphs).toBool();
+  bool renderBidirectionaly = state->getValue(ShowFieldGlyphs::RenderBidirectionaly).toBool();
+  bool renderGlphysBellowThreshold = state->getValue(ShowFieldGlyphs::RenderGlyphsBellowThreshold).toBool();
+  float threshold = state->getValue(ShowFieldGlyphs::Threshold).toDouble();
+
+  //sets feild location for consant feild data 1: node centered 2: edge centered 3: face centered 4: cell centered
```
Comment (dcwhite) [inline src/Modules/Visualization/ShowFieldGlyphs.cc]:
```suggestion
  //sets field location for constant field data 1: node centered 2: edge centered 3: face centered 4: cell centered
```

File: src/Modules/Visualization/ShowFieldGlyphs.cc (hunk 7 above)
Comment (dcwhite) [inline src/Modules/Visualization/ShowFieldGlyphs.cc]: I think there's a better way to do this--there ought to be an enum value you can query and then switch on, that tells you the data location. I will check what it is.

Comment (dcwhite) [reply src/Modules/Visualization/ShowFieldGlyphs.cc]: I'm thinking it's `field->vfield()->basis_order()`, but that might not quite be what you need. We'll merge as is (after you fix the `feild/field` typos) and we can revisit after.

File: src/Core/GeometryPrimitives/Vector.h [hunk 9]
```diff
@@ -110,6 +110,7 @@ class Vector
     inline Vector& operator-=(const Vector&);
     inline double normalize();
     inline double safe_normalize();
+    inline Vector getArbitraryTangent();
```
Comment (jessdtate) [inline src/Core/GeometryPrimitives/Vector.h]:
```suggestion
    inline Vector getArbitraryTangent() const;
```

Comment (jessdtate) [reply src/Core/GeometryPrimitives/Vector.h]: this is causing a build error

File: src/Modules/Visualization/ShowFieldGlyphs.cc [hunk 10]
```diff
@@ -342,161 +404,195 @@ void GlyphBuilder::renderVectors(
...
+  bool renderBidirectionaly = state->getValue(ShowFieldGlyphs::RenderBidirectionaly).toBool();
+  bool renderGlphysBellowThreshold = state->getValue(ShowFieldGlyphs::RenderGlyphsBellowThreshold).toBool();
+  float threshold = state->getValue(ShowFieldGlyphs::Threshold).toDouble();
+
+  //sets field location for constant field data 1: node centered 2: edge centered 3: face centered 4: cell centered
+  int fieldLocation = finfo.is_point()*1 + finfo.is_line()*2 + finfo.is_surface()*3 + finfo.is_volume()*4;
+  //sets feild location to 0 for linear data regardless of location
+  feildLocation = feildLocation * !finfo.is_linear();
+  
+  switch(feildLocation)
```
Comment (jessdtate) [inline src/Modules/Visualization/ShowFieldGlyphs.cc]:
```suggestion
  switch(fieldLocation)
```

Comment (jessdtate) [reply src/Modules/Visualization/ShowFieldGlyphs.cc]: this needs to be corrected in other places too

File: src/Modules/Visualization/ShowFieldGlyphs.cc [hunk 11]
```diff
@@ -342,161 +404,195 @@ void GlyphBuilder::renderVectors(
...
+  
+  bool normalizeGlyphs = state->getValue(ShowFieldGlyphs::NormalizeGlyphs).toBool();
+  bool renderBidirectionaly = state->getValue(ShowFieldGlyphs::RenderBidirectionaly).toBool();
+  bool renderGlphysBellowThreshold = state->getValue(ShowFieldGlyphs::RenderGlyphsBellowThreshold).toBool();
+  float threshold = state->getValue(ShowFieldGlyphs::Threshold).toDouble();
+
+  //sets field location for constant field data 1: node centered 2: edge centered 3: face centered 4: cell centered
+  int fieldLocation = finfo.is_point()*1 + finfo.is_line()*2 + finfo.is_surface()*3 + finfo.is_volume()*4;
+  //sets field location to 0 for linear data regardless of location
+  fieldLocation = feildLocation * !finfo.is_linear();
```
Comment (jessdtate) [inline src/Modules/Visualization/ShowFieldGlyphs.cc]:
```suggestion
  fieldLocation = fieldLocation * !finfo.is_linear();
```

**Pull request merged**